runtime/src/accounts_db.rs

```diff
@@ -3396,30 +3408,81 @@ impl AccountsDb {
-        // which version their parents because banks will also be augmented with this version,
-        // which handles cases where a deletion of one version happens in the middle of the scan.
+    // TODO: This is currently unsafe with scan because it can remove a slot in the middle
+    pub fn remove_unrooted_slots(&self, remove_slots: &[Slot]) {
+        let rooted_slots = self
+            .accounts_index
+            .get_rooted_from_list(remove_slots.iter());
+        if !rooted_slots.is_empty() {
+            panic!(
+                "Trying to remove accounts for rooted slots {:?}",
+                rooted_slots
+            );
+        }
+
+        let RemoveUnrootedSlots {
+            slots_under_contention,
+            signal,
+        } = &self.remove_unrooted_slots;
+
+        {
+            // Slots that are currently being flushed by flush_slot_cache()
+            let mut contended_slots = slots_under_contention.lock().unwrap();
+
+            // Slots that are currently being flushed by flush_slot_cache() AND
+            // we want to remove in this function
+            let mut contended_cache_flush_slots: Vec<Slot> = remove_slots
+                .iter()
+                .filter(|remove_slot| {
+                    let is_cache_flushing_slot = contended_slots.contains(remove_slot);
+                    if !is_cache_flushing_slot {
+                        // Reserve the slots that we want to purge that aren't currently
+                        // being flushed to prevent cache from flushing those slots in
+                        // the future
+                        contended_slots.insert(**remove_slot);
+                    }
+                    // If the cache is currently flushing this slot, add it to the list
+                    is_cache_flushing_slot
+                })
+                .cloned()
+                .collect();
+
+            // Wait for cache flushes to finish
+            loop {
+                if !contended_cache_flush_slots.is_empty() {
+                    // Wait for the signal that the cache has finished flushing a slot
+                    //
+                    // Don't wait if the contended_cache_flush_slots is empty, otherwise
+                    // we may never get a signal since there's no cache flush thread to
+                    // do the signaling
+                    contended_slots = signal.wait(contended_slots).unwrap();
+                } else {
+                    // There are no slots being flushed to wait on, so it's safe to continue
+                    // to purging the slots we want to purge!
+                    break;
+                }
+
+                // For each slot the cache flush has finished, mark that we're about to start
+                // purging these slots by reserving it in `contended_slots`.
+                contended_cache_flush_slots.retain(|flushing_slot| {
+                    let is_being_flushed = contended_slots.contains(flushing_slot);
```

Comment:
@lijunwangs, the `signal.wait(contended_slots).unwrap()` above releases the `contended_slots` and then regrabs the lock when the condition variable gets a signal, see documentation here: https://doc.rust-lang.org/std/sync/struct.Condvar.html#method.wait

>When it wakes up, how come the slot in `contended_cache_flush_slots` is not found in contended_slots
So the flush thread is the thread removing the slots from `contended_slots`, which is the shared variable under the lock. The `contended_cache_flush_slots` is a local snapshot of the slots in `contended_slots` at the time we checked it at the beginning of the function, and is never added to again after we filter from it, so it's always shrinking after each iteration of the `retain` above. We only add a slot to the `contended_slots` through`contended_slots.insert(*flushing_slot)` *after* we've confirmed the *retain* will remove the `slot ` because `!is_being_flushed` is true.

Reply:
Thanks explaining. Still have a question. In line 3445, contended_cache_flush_slots only has the slots which is in the also in contended_slots. Doesn't line 3471 re-add it again?

Reply:
ha yeah it does, I don't think it's strictly necessary, but I wanted to uphold the invariant that any slot that is *either* undergoing flush or `remove_unrooted_slots` are both present in `contended_slots`, so we always have the working set of slots observable by any thread.